### ISSUE TITLE
Solution loading perf has tanked.  Time to first edit is huge

### ISSUE THREAD

**Issue by CyrusNajmabadi** [dotnet/roslyn#20542]
See attached gif.  It looks like AsyncSolutionLoad is not working for Roslyn at all.  You have to load hte entire solution before you can do anything.  On my machine it took around 150 seconds (2.5 minutes) to get to that point.  

This feels like a huge regression.  I'm editing out of the dev15.6 branch.

**Comment by CyrusNajmabadi** [comment]
![slowload](https://user-images.githubusercontent.com/4564579/27699920-cb8f7ebc-5cb0-11e7-915b-64b2f5f5292d.gif)

**Comment by CyrusNajmabadi** [comment]
Note: the gif is double speed to not subject you to all the pain.

**Comment by CyrusNajmabadi** [comment]
Tagging @Pilchie Do you know what makes ASL stop working?

**Comment by jaredpar** [comment]
@CyrusNajmabadi couple Q's 

- What version of VS are you using? 
- What commit are you sync'd to? 

The commit is important as I just merged the new SDK switch this morning. That should actually improve solution load time (did on mine and @sharwell machine).

**Comment by CyrusNajmabadi** [comment]
i'm using 
![image](https://user-images.githubusercontent.com/4564579/27700025-2d0b0260-5cb1-11e7-894d-a86bf15a12a7.png)

I'm at commit: 331b72ba494f46fdeba17dae541d2b9aaeb91f99

The latest merge into this branch is 
29da0cb6be6aed21f5415d853851ee4af8a0544d (which is Sam's ending-newline change).  

When did your sdk change go in?  I can try to see if htat helps out.

**Comment by CyrusNajmabadi** [comment]
Ok.  I just fetched and merged in your change 8de366529ac0572921cd49d715dfcc370575856b   Will try this out now!

**Comment by CyrusNajmabadi** [comment]
(crossing fingers)

**Comment by jaredpar** [comment]
Note: I've tried the full SDK change with builds I'm on 26616.2 and 26621.2.

**Comment by CyrusNajmabadi** [comment]
Ok.  Your change helps a lot.  It drops things on my machine to just over one minute (150s to 60s!) That's AWESOME.  Note: asl still seems broken.  It looks like the entire solution needs to load before any editing can happen.

**Comment by CyrusNajmabadi** [comment]
Wait... i may have been wrong.  Now i'm getting UI hangs with this being shown:

![image](https://user-images.githubusercontent.com/4564579/27700443-72c9f148-5cb2-11e7-9d99-7986309323c9.png)

Is this related/unrelated?

**Comment by CyrusNajmabadi** [comment]
oh... vs is restoring nuget... wtf.  i told it not to do that...

**Comment by jaredpar** [comment]
Just about to ask if you disabled NuGet restore in the options page.

**Comment by CyrusNajmabadi** [comment]
Closing VS takes an eternity now.  Will upload gif.  Note: i can VS actually removing stuff from solution explorer while unloading the projects.  Why do we need to do any of that?  If i ask VS to close we shouldn't have to do any of this work right?  (Note: i'm closing VS, not closing a solution).

**Comment by CyrusNajmabadi** [comment]
![closevs](https://user-images.githubusercontent.com/4564579/27700821-b1f2c498-5cb3-11e7-8dda-0aaceae7d3cd.gif)

**Comment by CyrusNajmabadi** [comment]
It was so slow i could take the time to start up the encoder and then get back to VS.  It still had only made it to unloading the 6th project.  This was a multi-*MINUTE* shutdown.  Tagging @srivatsn @davkean

**Comment by CyrusNajmabadi** [comment]
During this time i could see projects being removed from the solution explorer pane.  

1. i have no idea why this would possibly be this slow.
2. i have no idea why we're doing any work on VS shutdown.

**Comment by CyrusNajmabadi** [comment]
Jared, these are my nuget options:

![image](https://user-images.githubusercontent.com/4564579/27701002-591d4c5c-5cb4-11e7-80e6-63afdf7fcc5e.png)

Do i need to uncheck the first guy as well?

**Comment by jaredpar** [comment]
I have them all unchecked on my machine.

**Comment by CyrusNajmabadi** [comment]
K.  will try that

**Comment by CyrusNajmabadi** [comment]
One thing i noticed was this:

![image](https://user-images.githubusercontent.com/4564579/27702875-64daf386-5cba-11e7-96bc-9e77ba289081.png)

It looks like during shutdown we're not cancelling whatever msbuild shenanigans are going on.  Maybe we're doing design time builds or something?  And it looks like we might be waiting on all that to end.  We need to cancel everything and do as little work as possible on shutdown.

**Comment by AdamSpeight2008** [comment]
Could it also be Live Unit Testing?

**Comment by Pilchie** [comment]
Sounds like this bug should be filed on `dotnet/project-system`.  @srivatsn @davkean ?

**Comment by srivatsn** [comment]
Is this about the closing issue now - if so https://github.com/dotnet/project-system/issues/2211 is tracking that.

**Comment by CyrusNajmabadi** [comment]
Ok thanks.  Let's keep this issue around for the broken ASL issue.  @Pilchie do you know who that would go to?

**Comment by Pilchie** [comment]
Either `dotnet/project-system` again, or the VS Platform solution team, which would need at VSTS bug.

**Comment by srivatsn** [comment]
It would be the platform team - we don't control ASL. Wasn't there an A\B experiment to turning ASL on\off for internal users? @IlyaBiryukov ?

**Comment by IlyaBiryukov** [comment]
Yes, there is an A/B experiment to turn off ASL. P&T team is still running it I suppose.

**Comment by davkean** [comment]
Does ASL kick in if you open again?

**Comment by IlyaBiryukov** [comment]
ASL mode is controlled by the flight of the A/B experiment. If the asloff flight is on, ASL will be off all the time.

**Comment by CyrusNajmabadi** [comment]
What's the purpose of this A/B test?  It seems to make time-to-edit incredibly slow.

**Comment by IlyaBiryukov** [comment]
Below are the results of the 1st week of the experiment with approximately 15K users on each flight:
* 95th percentile of the solution load duration in ASL-Off mode 8 seconds less than with ASL mode. Difference is statistically significant (validated used confidence interval).
* For the 75th percentile customers start working with solution approximately 13 seconds sooner for ASL-Off flight, than for ASLCF. Difference is statistically significant (validated used confidence interval).

I've started an email thread with Viktor and Pavel about this issue.

**Comment by CyrusNajmabadi** [comment]
Can you include me?   I want to make sure that we understand why it's better for some, but so bad for Roslyn.sln.  If, for example, we stopped using ASL it would be a huge regression in time-to-edit for people working on the Roslyn codebase.  Thanks!

**Comment by CyrusNajmabadi** [comment]
@jinujoseph This no longer seems to eb a problem.  If you want, feel free to close out.
